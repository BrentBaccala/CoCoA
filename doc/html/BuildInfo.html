<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">
<HTML>
<HEAD>
<META NAME="generator" CONTENT="http://txt2tags.sf.net">
<LINK REL="stylesheet" TYPE="text/css" HREF="cocoalib-doc.css">
<TITLE>      BuildInfo</TITLE>
</HEAD><BODY BGCOLOR="white" TEXT="black">
<P ALIGN="center"><CENTER><H1>      BuildInfo</H1>
<FONT SIZE="4">
<I>      &copy;  2005,2007 John Abbott</I><BR>
      GNU Free Documentation License, Version 1.2
</FONT></CENTER>

<P></P>
<HR NOSHADE SIZE=1>
<P></P>
    <UL>
    <LI><A HREF="#toc1">Examples</A>
    <LI><A HREF="#toc2">User documentation</A>
    <LI><A HREF="#toc3">Maintainer documentation</A>
    <LI><A HREF="#toc4">Bugs, Shortcomings and other ideas</A>
    </UL>

<P></P>
<HR NOSHADE SIZE=1>
<P></P>
<P>
      <center><a href="index.html">CoCoALib Documentation Index</a></center>
</P>
<A NAME="toc1"></A>
<H2>Examples</H2>
<UL>
<LI><A HREF="../../examples/index.html#ex-BuildInfo.C">ex-BuildInfo.C</A>
<LI><A HREF="../../examples/index.html#ex-limits.C">ex-limits.C</A>
</UL>

<P></P>
<A NAME="toc2"></A>
<H2>User documentation</H2>
<P>
The constant in <CODE>BuildInfo</CODE> allows you to find out which version of
CoCoALib you are using.  The function <CODE>BuildInfo::PrintAll</CODE> prints out all
the build information on the <CODE>ostream</CODE> passed in as argument -- you should
include this information whenever you report a bug.
</P>
<P>
There is one string constant which contains the version number of the
library: it is of the form <CODE>A.bcde</CODE> where <CODE>A</CODE> is the major
version, <CODE>bc</CODE> is the minor version, and <CODE>de</CODE> is the patch level.
Note that there are always precisely 4 digits after the point (even if
they are all zero).
</P>
<UL>
<LI><CODE>BuildInfo::version</CODE> -- a C string containing the CoCoALib version number.
</UL>

<P></P>
<P>
NOTE: if you happen upon a copy of libcocoa.a and want to find out which
version it is, you can use the following Unix/Linux command:
</P>
<PRE>
    strings libcocoa.a | egrep "CoCoA::BuildInfo"
</PRE>
<P></P>
<P>
This should print out three lines informing you of the library version,
the compiler used, and the compiler flags used when creating libcocoa.a.
</P>
<A NAME="toc3"></A>
<H2>Maintainer documentation</H2>
<P>
I chose to put the constants and function in their own namespace to
emphasise that they go together.
</P>
<P>
There are actually four string constants, but only one is supposed to
be publicly accessible (because I cannot imagine why anyone would want
access to the other three).  I made the constants C strings because it
seemed simpler than using C++ strings.  The three constants
<CODE>VersionMesg</CODE>, <CODE>CompilerMesg</CODE>, and <CODE>CompilerFlagsMesg</CODE> contain
the substring <CODE>CoCoA::BuildInfo</CODE> so that the egrep trick described
above will produce the version/compiler information directly.
</P>
<P>
I made <CODE>BuildInfo::PrintAll</CODE> leave a blank line before and after its
message so that it would stand out better from other output produced
by the program.
</P>
<A NAME="toc4"></A>
<H2>Bugs, Shortcomings and other ideas</H2>
<P>
The constants are not C++ strings -- is this really a bug?
</P>
<P>
Should the three constants <CODE>VersionMesg</CODE>, <CODE>CompilerMesg</CODE>, and
[CompilerFlagsMesg] be hidden or public?  Until someone convinces me there
is a good reason to make them public, they'll stay private.
</P>

<!-- html code generated by txt2tags 2.5 (http://txt2tags.sf.net) -->
<!-- cmdline: txt2tags -t html -o html/BuildInfo.html -i txt/BuildInfo.txt -->
</BODY></HTML>
