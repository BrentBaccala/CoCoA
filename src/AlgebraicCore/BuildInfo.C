//   Copyright (c)  2005,2007,2012  John Abbott

//   This file is part of the source of CoCoALib, the CoCoA Library.

//   CoCoALib is free software: you can redistribute it and/or modify
//   it under the terms of the GNU General Public License as published by
//   the Free Software Foundation, either version 3 of the License, or
//   (at your option) any later version.

//   CoCoALib is distributed in the hope that it will be useful,
//   but WITHOUT ANY WARRANTY; without even the implied warranty of
//   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//   GNU General Public License for more details.

//   You should have received a copy of the GNU General Public License
//   along with CoCoALib.  If not, see <http://www.gnu.org/licenses/>.


#include "CoCoA/BuildInfo.H"
#include "CoCoA/config.H"

#include <limits>
#include <iostream>
using std::ostream;
using std::endl;
#include <string>
using std::string;

namespace CoCoA
{

  namespace BuildInfo
  {
    // Build info is actually passed in via three preprocessor variables:
    // COCOA_VERSION, COCOA_CXX, COCOA_CXXFLAGS -- all three are quoted strings.

    const std::string& version()
    {
      static const string info(COCOA_VERSION);
      return info;
    }

    const std::string& compiler()
    {
      static const string info(COCOA_CXX);
      return info;
    }

    const std::string& CompilationFlags()
    {
      static const string info(COCOA_CXXFLAGS);
      return info;
    }

    const std::string& CompilationDefines()
    {
      static const string info(COCOA_CXXFLAGS_DEFINES);
      return info;
    }


    void PrintAll(std::ostream& out)
    {
    // Below we use string literal juxtaposition, so that the build
    // info can readily be extracted from libcocoa.a by the command
    // strings libcocoa.a | egrep "CoCoA::BuildInfo"
      out << endl
          << "CoCoA::BuildInfo Summary of build information for CoCoALib:" << endl
          << "CoCoA::BuildInfo CoCoALib Version: " COCOA_VERSION << endl
          << "CoCoA::BuildInfo Compiler: " COCOA_CXX << endl
          << "CoCoA::BuildInfo Compilation Flags: " COCOA_CXXFLAGS << endl
          << "CoCoA::BuildInfo Compilation Flags Defines: " COCOA_CXXFLAGS_DEFINES << endl
          << "CoCoA::BuildInfo NumBits int = " << std::numeric_limits<unsigned int>::digits << endl
          << "CoCoA::BuildInfo NumBits long = " << std::numeric_limits<unsigned long>::digits << endl
          << "CoCoA::BuildInfo NumBits SmallExponent_t = " << std::numeric_limits<SmallExponent_t>::digits << endl
          << endl;
    }

  } // end of namespace CoCoA::BuildInfo

} // end of namespace CoCoA


// RCS header/log in the next few lines
// $Header: /Volumes/Home_1/cocoa/cvs-repository/CoCoALib-0.99/src/AlgebraicCore/BuildInfo.C,v 1.8 2012/10/24 12:10:45 abbott Exp $
// $Log: BuildInfo.C,v $
// Revision 1.8  2012/10/24 12:10:45  abbott
// Revised PrintAll: now prints also NumBits of SmallExponent_t
//
// Revision 1.7  2012/10/02 11:45:09  bigatti
// -- added CompilationDefines
//
// Revision 1.6  2012/10/02 10:36:12  abbott
// Revised interface to BuildInfo information strings.
// Several consequential changes.
//
// Revision 1.5  2012/06/20 12:26:26  bigatti
// -- now exporting also compiler and compilation flags
//
// Revision 1.4  2011/03/11 17:38:21  bigatti
// -- added missing include
//
// Revision 1.3  2011/03/11 17:18:55  bigatti
// -- added info on int and long size
//
// Revision 1.2  2007/10/30 17:14:08  abbott
// Changed licence from GPL-2 only to GPL-3 or later.
// New version for such an important change.
//
// Revision 1.1.1.1  2007/03/09 15:16:11  abbott
// Imported files
//
// Revision 1.4  2007/02/10 18:44:03  cocoa
// Added "const" twice to each test and example.
// Eliminated dependency on io.H in several files.
// Improved BuildInfo, and added an example about how to use it.
// Some other minor cleaning.
//
// Revision 1.3  2007/02/08 22:34:22  cocoa
// Changed BuildInfo: only BuildInfo version is publicly visible.
// Only BuildInfo needs complicated compile-time flags, so several changes
// to Makefiles etc.  Added a new example: ex-BuildInfo.C
//
// Revision 1.1.1.1  2005/10/17 10:46:54  cocoa
// Imported files
//
// Revision 1.2  2005/10/12 15:52:09  cocoa
// Completed test-RingFp1 and corrected/cleaned the SmallFp*
// and RingFp* files.
//
// Some minor tidying elsewhere.
//
// Revision 1.1  2005/10/06 16:36:42  cocoa
// Added the capability find out build information at run-time.
// The Makefiles should be a little tidier too.
//
