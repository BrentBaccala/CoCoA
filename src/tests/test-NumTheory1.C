//   Copyright (c)  2009-2010  John Abbott

//   This file is part of the source of CoCoALib, the CoCoA Library.

//   CoCoALib is free software: you can redistribute it and/or modify
//   it under the terms of the GNU General Public License as published by
//   the Free Software Foundation, either version 3 of the License, or
//   (at your option) any later version.

//   CoCoALib is distributed in the hope that it will be useful,
//   but WITHOUT ANY WARRANTY; without even the implied warranty of
//   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//   GNU General Public License for more details.

//   You should have received a copy of the GNU General Public License
//   along with CoCoALib.  If not, see <http://www.gnu.org/licenses/>.


#include "CoCoA/BuildInfo.H"
#include "CoCoA/error.H"
#include "CoCoA/VectorOperations.H"
#include "CoCoA/GlobalManager.H"
#include "CoCoA/NumTheory.H"
#include "CoCoA/IntOperations.H"

#include <cmath>
using std::log;
#include <cstdlib>
using std::abs;
#include <iomanip>
using std::flush;
#include <iostream>
using std::cout;
using std::cerr;
using std::endl;
#include <limits>
using std::numeric_limits;
#include <vector>
using std::vector;

using namespace CoCoA;

#define TEST_ASSERT(cond)   do { if (!(cond)) CoCoA::AssertionFailed(#cond, __FILE__, __LINE__); } while(0)




void program()
{
  GlobalManager CoCoAFoundations(UseGMPAllocator);

  cout << "Testing gcd and lcm on small numbers." << endl;
  const int Nmax = 257;
  for (int a=-Nmax; a < Nmax; ++a)
    for (int b=-Nmax; b < Nmax; ++b)
    {
      const BigInt A(a);
      const BigInt B(b);
      const int gcdab = gcd(a,b);
      TEST_ASSERT(gcdab >= 0);
      TEST_ASSERT((a == 0 && b == 0) || gcdab > 0);
      TEST_ASSERT(gcdab == 0 || (a%gcdab == 0 && b%gcdab == 0));
      TEST_ASSERT(gcdab == gcd(a,B));
      TEST_ASSERT(gcdab == gcd(A,b));
      TEST_ASSERT(gcdab == gcd(A,B));

      const int lcmab = lcm(a,b);
      TEST_ASSERT(gcdab*lcmab == abs(a*b));
      TEST_ASSERT(lcmab == lcm(a,B));
      TEST_ASSERT(lcmab == lcm(A,b));
      TEST_ASSERT(lcmab == lcm(A,B));
    }

  cout << "\nTesting ExtGcd on small numbers" << endl;
  for (int a=-Nmax; a < Nmax; ++a)
    for (int b=-Nmax; b < Nmax; ++b)
    {
      if (a == 0 && b == 0) continue;
      long CofacA, CofacB;
      const int g = ExtGcd(CofacA,CofacB,a,b);
      const int g2 = gcd(a,b);
      TEST_ASSERT(g == g2);
      TEST_ASSERT(CofacA*a+CofacB*b == g);
      if (abs(a) == abs(b))
      {
        TEST_ASSERT((CofacA == 0 && CofacB == sign(b)) || (CofacB == 0 && CofacA == sign(a)));
        continue;
      }
      if (abs(a) == g) { TEST_ASSERT(CofacA == sign(a) && CofacB == 0); continue; }
      if (abs(b) == g) { TEST_ASSERT(CofacB == sign(b) && CofacA == 0); continue; }
      // General case
      TEST_ASSERT(abs(CofacA) <= (abs(b)/g)/2  && abs(CofacB) <= (abs(a)/g)/2);
      if (2*g == abs(b))
        TEST_ASSERT(abs(CofacA) == 1);
      else
        TEST_ASSERT(2*abs(CofacA) < abs(b)/g);
      if (2*g == abs(a))
        TEST_ASSERT(abs(CofacB) == 1);
      else
        TEST_ASSERT(2*abs(CofacB) < abs(a)/g);
    }


  cout << "\nTesting ExtGcd on BigInt numbers" << endl;
  for (int a=-Nmax; a < Nmax; ++a)
    for (int b=-Nmax; b < Nmax; ++b)
    {
      if (a == 0 && b == 0) continue;
      BigInt A(a);
      BigInt B(b);
      BigInt CofacA, CofacB;
      const BigInt G = ExtGcd(CofacA,CofacB,A,B);
      const int g = gcd(a,b);
      TEST_ASSERT(G == g);
      TEST_ASSERT(CofacA*a+CofacB*b == g);
      if (abs(a) == abs(b))
      {
        TEST_ASSERT((CofacA == 0 && CofacB == sign(b)) || (CofacB == 0 && CofacA == sign(a)));
        continue;
      }
      if (abs(a) == g) { TEST_ASSERT(CofacA == sign(a) && CofacB == 0); continue; }
      if (abs(b) == g) { TEST_ASSERT(CofacB == sign(b) && CofacA == 0); continue; }
      // General case
      TEST_ASSERT(abs(CofacA) <= (abs(b)/g)/2  && abs(CofacB) <= (abs(a)/g)/2);
      if (2*g == abs(b))
        TEST_ASSERT(abs(CofacA) == 1);
      else
        TEST_ASSERT(2*abs(CofacA) < abs(b)/g);
      if (2*g == abs(a))
        TEST_ASSERT(abs(CofacB) == 1);
      else
        TEST_ASSERT(2*abs(CofacB) < abs(a)/g);
    }


  cout << "\nTesting InvMod" << endl;
  for (int r=-Nmax; r < Nmax; ++r)
    for (int m=2; m < Nmax; ++m)
    {
      BigInt R(r);
      BigInt M(m);
      const int invr = InvMod(r,m);
      TEST_ASSERT(gcd(r,m)==1 || invr==0);
      if (invr==0) continue;
      if (r > 0) TEST_ASSERT((r*invr)%m == 1);
      if (r < 0) TEST_ASSERT(((-r)*invr)%m == m-1);
      TEST_ASSERT(invr == InvMod(R,m));
      TEST_ASSERT(invr == InvMod(r,M));
      TEST_ASSERT(invr == InvMod(R,M));
    }

  cout << "\nTesting PowerMod" << endl;
  for (int r=-Nmax; r < Nmax; ++r)
    for (int m=2; m < Nmax; ++m)
    {
      BigInt R(r);
      BigInt M(m);
      TEST_ASSERT(PowerMod(r,0,m) == 1);
      TEST_ASSERT(PowerMod(r,0,M) == 1);
      TEST_ASSERT(PowerMod(R,0,m) == 1);
      TEST_ASSERT(PowerMod(R,0,M) == 1);

      long RmodM;
      if (r >= 0) RmodM = r%m; else RmodM = (m-((-r)%m))%m;
      TEST_ASSERT(PowerMod(r,1,m) == RmodM);
      TEST_ASSERT(PowerMod(r,1,M) == RmodM);
      TEST_ASSERT(PowerMod(R,1,m) == RmodM);
      TEST_ASSERT(PowerMod(R,1,M) == RmodM);

      if (gcd(r,m) == 1)
      {
        long InvR = InvMod(r,m);
        TEST_ASSERT(PowerMod(r,-1,m) == InvR);
        TEST_ASSERT(PowerMod(r,-1,M) == InvR);
        TEST_ASSERT(PowerMod(R,-1,m) == InvR);
        TEST_ASSERT(PowerMod(R,-1,M) == InvR);
      }
      for (int e=0; e < 9; ++e)
      {
        TEST_ASSERT(PowerMod(PowerMod(r,e,m), 2, m) == PowerMod(r, 2*e, m));
        if (gcd(r,m) == 1)
          TEST_ASSERT((PowerMod(r,e,m)*PowerMod(r,-e,m))%m == 1);
      }
    }



  cout << "\nTesting IsPrime for small numbers (incl negative ones)." << endl;
  int PrimeCounter = 0;
  for (int n=1; n <= 1000; ++n)
  {
    if (IsPrime(n)) ++PrimeCounter;
    TEST_ASSERT(IsPrime(n) == IsPrime(BigInt(n)));
  }
  TEST_ASSERT(PrimeCounter == 168);

  // ATTN!  this next test depends on the platform (32-bit or 64-bit)
  const int LongBits = numeric_limits<long>::digits;
  cout << "\nTesting IsPrime for numbers just greater than max long." << endl;
  const BigInt StartingPoint = power(2,LongBits);
  for (int delta=1; delta < 99; delta+=2)
  {
    const BigInt N = StartingPoint+delta;
    if (!IsProbPrime(N)) continue;
    if (IsPrime(N)) continue;
    cout << "Surprise: " << N << " is a composite pseudo-prime." << endl;
  }


  cout << "\nTesting IsProbPrime: lengths of repunit (probable) primes (up to 500):" << flush;
  for (int n=1; n <= 500; ++n)
  {
    if (!IsPrime(n)) continue; // no need to try composite n
    if (IsProbPrime((power(10,n)-1)/9))
      cout << " " << n << flush;
  }
  cout << endl;


  cout << "\nTesting NextPrime & PrevPrime" << endl;
  const long MaxLong = numeric_limits<long>::max();
  TEST_ASSERT(NextPrime(1) == 2);
  TEST_ASSERT(NextPrime(2) == 3);
  TEST_ASSERT(NextPrime(3) == 5);
  TEST_ASSERT(NextPrime(4) == 5);

  TEST_ASSERT(NextPrime(MaxLong) == 0);
  for (long n=MaxLong; !IsPrime(n--); )
    TEST_ASSERT(NextPrime(n) == 0);

  const long MaxPrime = IsPrime(MaxLong)?MaxLong:PrevPrime(MaxLong);
  TEST_ASSERT(IsPrime(MaxPrime));
  TEST_ASSERT(NextPrime(MaxPrime) == 0);
  for (long n=MaxPrime; !IsPrime(n--); )
    TEST_ASSERT(NextPrime(n) == MaxPrime);

  TEST_ASSERT(PrevPrime(4) == 3);
  TEST_ASSERT(PrevPrime(3) == 2);
  TEST_ASSERT(PrevPrime(2) == 0);
  TEST_ASSERT(PrevPrime(1) == 0);
  for (int n=3; n < 1000; n += 2)
  {
    if (IsPrime(n))
    {
      TEST_ASSERT(PrevPrime(NextPrime(n)) == n);
      TEST_ASSERT(NextPrime(PrevPrime(n)) == n);
    }
    else
    {
      TEST_ASSERT(NextPrime(PrevPrime(n)) == NextPrime(n));
      TEST_ASSERT(PrevPrime(NextPrime(n)) == PrevPrime(n));
    }
  }

  cout << "\nTesting PrimitiveRoot" << endl;
  const int MaxHist = 20;
  vector<int> histogram(MaxHist);
  int p = 1;
  while (p < 65535)
  {
    p = NextPrime(p);
    const int g = PrimitiveRoot(p);
    if (g < MaxHist) ++histogram[g];
//     double ratio = g/(std::log(p)*std::log(std::log(p)));
//     if (ratio > 2.0) cout << "ratio=" << ratio << "  for p=" << p << " and g=" << g << endl;
  }
  cout << "Histogram of least positive primitive roots: " << histogram << endl;


  cout << "\nTesting factorize on smaller & larger integers:" << endl;
  for (int i=1; i <= 16; ++i)
  {
    const BigInt N = power(10,i)-1;
    cout << "Factorization of " << N << ": " << factor(N) << endl;
  }

  cout << "\nTesting EulerPhi on small numbers." << endl;
  cout << "Here are the most frequent values of EulerPhi on numbers up to 100000:\n"
                 << "Value  Freq\n";
  vector<int> freq(100000);
  for (int i=2; i < 100000; ++i)
    ++freq[EulerPhi(i)];
  for (int i=2; i < 100000; ++i)
    if (freq[i] > 200) 
      cout << i << "  " << freq[i] << endl;

  cout << "\nTesting EulerPhi on larger numbers." << endl
                 << "Looking for smallest number with EulerPhi(N) < N/10." << endl;
  BigInt BigNum;
  BigNum = 2;
  int LastPrime = -1;
  for (int p=3; 10*EulerPhi(BigNum) > BigNum; p=NextPrime(p))
  {
    BigNum *= p;
    LastPrime = p;
  }
  BigInt PhiBigNum = EulerPhi(BigNum);
  cout << "Result is  N=" << BigNum << endl
                 << "Its EulerPhi=" << PhiBigNum << endl
                 << "In fact N is just the product of all primes up to " << LastPrime << endl;


  // SmoothFactor
  cout << "\nTesting SmoothFactor" << endl;
  factorization<BigInt> FactoredFactorial = SmoothFactor(factorial(101), 101);
  cout << "101-smooth factors of factorial of 101: " << FactoredFactorial << endl;
  FactoredFactorial = SmoothFactor(factorial(101), 50);
  cout << "50-smooth factors of factorial of 101: " << FactoredFactorial << endl;
}


// Use main() to handle any uncaught exceptions and warn the user about them.
int main()
{
  try
  {
    program();
    return 0;
  }
  catch (const CoCoA::ErrorInfo& err)
  {
    cerr << "***ERROR***  UNCAUGHT CoCoA Error";
    ANNOUNCE(cerr, err);
  }
  catch (const std::exception& exc)
  {
    cerr << "***ERROR***  UNCAUGHT std::exception: " << exc.what() << endl;
  }
  catch(...)
  {
    cerr << "***ERROR***  UNCAUGHT UNKNOWN EXCEPTION" << endl;
  }

  BuildInfo::PrintAll(cerr);
  return 1;
}
