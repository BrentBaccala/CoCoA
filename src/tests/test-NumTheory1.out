Testing gcd and lcm on small numbers.

Testing ExtGcd on small numbers

Testing ExtGcd on BigInt numbers

Testing InvMod

Testing PowerMod

Testing IsPrime for small numbers (incl negative ones).

Testing IsPrime for numbers just greater than max long.

Testing IsProbPrime: lengths of repunit (probable) primes (up to 500): 2 19 23 317

Testing NextPrime & PrevPrime

Testing PrimitiveRoot
Histogram of least positive primitive roots: [0,  1,  2443,  1520,  0,  932,  370,  466,  0,  0,  140,  237,  23,  154,  42,  23,  0,  70,  1,  46]

Testing factorize on smaller & larger integers:
Factorization of 9: factorization(myFactors=[3], myMultiplicities=[2], myRemainingFactor=1)
Factorization of 99: factorization(myFactors=[3,  11], myMultiplicities=[2,  1], myRemainingFactor=1)
Factorization of 999: factorization(myFactors=[3,  37], myMultiplicities=[3,  1], myRemainingFactor=1)
Factorization of 9999: factorization(myFactors=[3,  11,  101], myMultiplicities=[2,  1,  1], myRemainingFactor=1)
Factorization of 99999: factorization(myFactors=[3,  41,  271], myMultiplicities=[2,  1,  1], myRemainingFactor=1)
Factorization of 999999: factorization(myFactors=[3,  7,  11,  13,  37], myMultiplicities=[3,  1,  1,  1,  1], myRemainingFactor=1)
Factorization of 9999999: factorization(myFactors=[3,  239,  4649], myMultiplicities=[2,  1,  1], myRemainingFactor=1)
Factorization of 99999999: factorization(myFactors=[3,  11,  73,  101,  137], myMultiplicities=[2,  1,  1,  1,  1], myRemainingFactor=1)
Factorization of 999999999: factorization(myFactors=[3,  37,  333667], myMultiplicities=[4,  1,  1], myRemainingFactor=1)
Factorization of 9999999999: factorization(myFactors=[3,  11,  41,  271,  9091], myMultiplicities=[2,  1,  1,  1,  1], myRemainingFactor=1)
Factorization of 99999999999: factorization(myFactors=[3,  21649,  513239], myMultiplicities=[2,  1,  1], myRemainingFactor=1)
Factorization of 999999999999: factorization(myFactors=[3,  7,  11,  13,  37,  101,  9901], myMultiplicities=[3,  1,  1,  1,  1,  1,  1], myRemainingFactor=1)
Factorization of 9999999999999: factorization(myFactors=[3,  53,  79,  265371653], myMultiplicities=[2,  1,  1,  1], myRemainingFactor=1)
Factorization of 99999999999999: factorization(myFactors=[3,  11,  239,  4649,  909091], myMultiplicities=[2,  1,  1,  1,  1], myRemainingFactor=1)
Factorization of 999999999999999: factorization(myFactors=[3,  31,  37,  41,  271,  2906161], myMultiplicities=[3,  1,  1,  1,  1,  1], myRemainingFactor=1)
Factorization of 9999999999999999: factorization(myFactors=[3,  11,  17,  73,  101,  137,  5882353], myMultiplicities=[2,  1,  1,  1,  1,  1,  1], myRemainingFactor=1)

Testing EulerPhi on small numbers.
Here are the most frequent values of EulerPhi on numbers up to 100000:
Value  Freq
17280  247
20160  244
23040  220
25920  238
30240  217

Testing EulerPhi on larger numbers.
Looking for smallest number with EulerPhi(N) < N/10.
Result is  N=16516447045902521732188973253623425320896207954043566485360902980990824644545340710198976591011245999110
Its EulerPhi=1651030565452496906032436282048608195577074117480554374969005278877765244397398104473600000000000000000
In fact N is just the product of all primes up to 257

Testing SmoothFactor
101-smooth factors of factorial of 101: factorization(myFactors=[2,  3,  5,  7,  11,  13,  17,  19,  23,  29,  31,  37,  41,  43,  47,  53,  59,  61,  67,  71,  73,  79,  83,  89,  97,  101], myMultiplicities=[97,  48,  24,  16,  9,  7,  5,  5,  4,  3,  3,  2,  2,  2,  2,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1], myRemainingFactor=1)
50-smooth factors of factorial of 101: factorization(myFactors=[2,  3,  5,  7,  11,  13,  17,  19,  23,  29,  31,  37,  41,  43,  47], myMultiplicities=[97,  48,  24,  16,  9,  7,  5,  5,  4,  3,  3,  2,  2,  2,  2], myRemainingFactor=378705860712501179527)
